go/vttest: Implement natively

vmg opened the pull request:
:wave: Hiii @sougou! I think @arthurnn briefly discussed this on Slack, but wouldn't it be cool if `vttest` didn't depend on a Python environment for spawning local test clusters?

I think that'd be N E A T, so this PR does that. Here's the commit message inlined with the details:

```
The vttest package is used by Go code (both internally and outside the
vitess project) as a way to configure and spawn a local self-contained
Vitess cluster that can be used for testing the service's behavior
without having to launch a fleet of hosts.

The self-contained cluster is composed of several processes (most
notably: a mysqld instance and a vtcombo service, which contains the
vtgate and vttablet parts of a normal Vitess deployment). These external
processes are currently controlled by `run_local_server.py`, a Python
script that configures and manages the deployment.

The current version of the `vttest` package simply keeps a
`run_local_server.py` process running in the background and reads
metadata from the cluster as JSON through stdout. This makes using
`vttest` for testing Go services a bit of a convoluted process, as you'd
require a full Python environment + all the dependencies of the Python
script (Python Protobufs + GRPC) just to control the service.
Introspection is also difficult as the information which can be queried
from the cluster's status is limited to the initial JSON payload that
`run_local_server.py` prints on startup.

To improve the situation of local testing through `vttest`, this commit
reimplements all the process-management logic from the original script
as native Go code.

The core of this implementation is the LocalCluster struct, the Go
equivalent to Python's LocalDatabase. This struct can be configured in
depth and then used to spawn and manage the services of the Vitess
cluster.

Here are some of the technical choices made during the rewrite:

- The existing `LaunchVitess(VitessOption...) *Handle` interface has
been kept intact for backwards compatibility reasons. Instead of
spawning the Python process, it now acts as a thin wrapper over
`LocalCluster`. The `VitessOption` calls are likewise used to set
configuration options on the `LocalCluster` struct, so overall the
external behavior of the service should remain 100% identical. All these
public APIs, however, have been marked as deprecated on their
documentation. The new `LocalCluster` interface is much more ergonomic,
and as such new users should ideally use it directly.

- The only exception to this backwards-compatibility is the `InitData()`
call: this option callback to setup random initialization of databases
was _not_ working at all, as the `InitDataOptions` struct that it
requires has no exported fields, and hence couldn't be used outside of
the library to configure the cluster.

- The `InitData` functionality has been exposed through the
`LocalCluster` interface with a properly exported `SeedConfig` struct.
However: the implementation for random initialization has not been fully
completed (randomized fields are not implemented) as it's not currently
used anywhere.

- The initialization and control of the `mysqld` service is performed
through an the `mysqlctl` command. `mysqlctl` is both a Go binary and a
Go package in the Vitess source tree; since this new `vttest`
implementation is fully native, it could in theory be possibly to import
the `go/vt/mysqlctl` package directly from Go and use it to initialize
the cluster. We have opted against this because the `mysqlctl` package
is principally configured through global flags, and hence not
threadsafe. By spawning a separate `mysqlctl` binary, we can ensure that
any Go test suite can use several `LocalCluster`s concurrently from many
goroutines without polluting the global namespace.

- With the same rationale, special care has been taken to ensure that
all the operations in `LocalCluster` are thread-safe and do not pollute
the global environment. Most notably, the `vttest` package doesn't
create any global flags or set any environment variables that could race
or break other concurrent tests. The environment variables for all
spawned processes are passed directly through `*spawn` and not inherited
from the parent.
```

There are a few things I'd like to follow up on, namely:

- [ ] Fix the `go/vt/mysqlctl` package so it's threadsafe and doesn't depend on global flags
- [ ] Change the `Mysqlctl` manager in `vttest` to use the `mysqlctl` package directly
- [ ] Design a more advanced interface in `SeedConfig` to configure how we're actually generating the random values, and support more kinds of column types for data generation (the Python script can only generate numerical columns right now!)
- [ ] Write a `go/cmd/vtlocalcluster` equivalent to `run_local_server.py`. Ideally it would be compatible with the Python script and a drop-in replacement! A statically linked binary for spawning the cluster would be very cool for projects that use Vitess and are not written in either Go or Python.

That's a long TODO list! I'll get there soon, but I'd appreciate if you could take a look at the current state of the PR and let me know whether it's going in the right direction.

Cheers!

alainjobart commented:
This is indeed neat. Thanks for doing this.

However, this won't work as is for us (but it's close!). Let me give you a bit of background:
- in our build environment, we use a custom mysqld, and we don't launch it with mysqlctl. It can take a list of extra my.cnf files, and just runs an instance.
- we also don't use random ports, but ports picked by our test environment.
- so by convention, we pick which class to use, which path to use, how to find ports in environment.{py,go}. Everytime you see one of these environment files, you can be sure we don't import that file, but we have a custom one internally that has to be manually updated every time the original one changes.
- that's why part of the init code is in py/vttest/mysql_db_mysqlctl.py and mysql_db_class exists. Internally, we have another class, and it's used instead.
- that is also why environment.{py,go} only contain a few configs and factory method, but no real code, so we don't maintain two versions of the code.

So I think in this PR, we'd need to abstract the mysqlctl stuff a bit more, and have a factory method in environment.go that creates a Mysqlctl object. Then internally we can implement our own MySQLManager, and just change the factory method in our environment.go. What you have added now in environment.go should probably go to another file too, as some of it is common.

Does that make sense?

arthurnn commented:
> in our build environment, we use a custom mysqld, and we don't launch it with mysqlctl. It can take a list of extra my.cnf files, and just runs an instance.

Totally make sense.. @vmg and I had that discussion earlier today.
In our setup, we also have MySQL already running and do not need to spin it with mysqlctl.
I am working to integrate this with our environment, @vmg and I will keep updating here as well.
Thanks

vmg commented:
@alainjobart: Thanks for that feedback! I knew something was off in that `environment.py` file (particularly the hardcoded "grpc" constant) -- I should have assumed you were replacing it inhouse!

Let's go for take two: I've implemented `Environment` as an interface that can be passed to each `LocalCluster` in a test suite. The API is as follows:

```go
// Environment is the interface that customizes the global settings for
// the test cluster. Usually the same environment settings are shared by
// all the LocalCluster instances in a given test suite, with each instance
// receiving a different Config for specific tests.
// For Environments that create temporary data on-disk and clean it up on
// termination, a brand new instance of Environment should be passed to
// each LocalCluster.
type Environment interface {
	// BinaryPath returns the full path to the given executable
	BinaryPath(bin string) string

	// MySQLManager is the constructor for the MySQL manager that will
	// be used by the cluster. The manager must take care of initializing
	// and destructing the MySQL instance(s) that will be used by the cluster.
	// See: vttest.MySQLManager for the interface the manager must implement
	MySQLManager(mycnf []string, snapshot string) (MySQLManager, error)

	// Directory is the path where the local cluster will store all its
	// data and metadata. For local testing, this should probably be an
	// unique temporary directory.
	Directory() string

	// LogDirectory is the directory where logs for all services in the
	// cluster will be stored.
	LogDirectory() string

	// VtcomoboArguments are the extra commandline arguments that will be
	// passed to `vtcombo`
	VtcomboArguments() []string

	// ProcessHealthCheck returns a HealthChecker for the given service.
	// The HealthChecker takes an address and attempts to check whether
	// the service is up and healthy.
	// If a given service does not require any custom health checks,
	// nil can be returned.
	ProcessHealthCheck(name string) HealthChecker

	// DefaultProtocol is the protocol used to communicate with the
	// Vitess cluster. This is usually "grpc".
	DefaultProtocol() string

	// PortForProtocol returns the listening port for a given service
	// on the given protocol. If protocol is empty, the default protocol
	// for each service is assumed.
	PortForProtocol(name, protocol string) int

	// EnvVars returns the environment variables that will be passed
	// to all Vitess processes spawned by the local cluster. These variables
	// always take precedence over the variables inherited from the current
	// process.
	EnvVars() []string

	// TearDown is called during LocalCluster.TearDown() to cleanup
	// any temporary data in the environment. Environments that can
	// last through several test runs do not need to implement it.
	TearDown() error
}
```

If you don't pass an explicit `Environment` to a `LocalCluster`, it'll get initialized with a default `LocalTestEnv`:

```go
// NewLocalTestEnv returns an instance of the default test environment used
// for local testing Vitess. The defaults are as follows:
// - Directory() is a random temporary directory in VTDATAROOT, which is cleaned
// up when closing the Environment.
// - LogDirectory() is the `logs` subdir inside Directory()
// - The MySQL flavor is set to `flavor`. If the argument is not set, it will
// default to the value of MYSQL_FLAVOR, and if this variable is not set, to
// DefaultMySQLFlavor
// - PortForProtocol() will return ports based off the given basePort. If basePort
// is zero, a random port between 10000 and 20000 will be chosen.
// - DefaultProtocol() is always "grpc"
// - ProcessHealthCheck() performs no service-specific health checks
// - BinaryPath() will look up the default Vitess binaries in VTROOT
// - MySQLManager() will return a vttest.Mysqlctl instance, configured with the
// given MySQL flavor. This will use the `mysqlctl` command to initialize and
// teardown a single mysqld instance.
func NewLocalTestEnv(flavor string, basePort int) (*LocalTestEnv, error)
```

As you can see, I combed through `run_local_server.py` and brought back all the abstractions which I had removed as they seemed no-ops. More specifically:

- It should now be possible to swap GRPC with Stubby, both as default protocol and by passing extra arguments to `vtcombo` to configure its RPC.
- All ports are customizable and do not need to depend off a base port
- Service-specific health checks can be implemented
- You can construct your own `MySQLManager` and initialize `mysqld` in whatever way
- You also get the list of custom CNF files when constructing the manager (on top of whatever flavor-specific CNFs are the default)
- The logs directory is customizable

I think those are all the abstractions in the original Python sources! Let me know if I missed any and I'll refactor this a lil' bit more 👌

sougou commented:
I've finally cleared my backlog and starting to look at this. It'll take me sometime to ramp up; I'll also need to build the adapters for it internally.

While I do that, can you take a look at Travis? One of the tests is failing.

vmg commented:
@sougou It seems like the failing test is a one-off thing, given that the tests pass for all other shard counts. Could you please retrigger the build and see if that fixes it?

michael-berlin commented:
I retriggered the tests. They are still failing.

If you have Docker installed, you can run them yourselves:

`./test.go unit` in the root directory

vmg commented:
@michael-berlin Ayup, apologies, I should have looked at the errors more carefully. They're now fixed. 👌

@sougou:

> Before I looked at the code, I thought we'd only need to replace mysqlctl.go and just reuse environment,go. If this is still possible, we could get rid of the Environment abstraction layer. Do you think this is still possible, or worth it?

The new `vttest` API has been designed so you _really_ shouldn't need to "replace" files. All the customizations in the old `run_local_cluster.py` should be replicable by injecting your own `Environment` implementation. We're doing that for our local tests and it works great. I would very much prefer to leave the abstraction in, as we don't vendor a custom version of Vitess and hence replacing the whole file would not work for us.

sougou commented:
Somehow, I thought the injection approach wouldn't work. I'll give it a second look.

alainjobart commented:
I'd agree with @vmg and try to add more implementations instead of replacing files. We just add go files internally, right? Is that what you guys do, add new internal-only files in the package?

Also you can leave the mysqlctl based implementation in there, no need to remove it, it can just be unused.

sougou commented:
I think the factory method is what's missing.

In your case, I believe you've written new tests that use this framework and they are capable of initializing LocalCluster differently. In our case, the existing tests (like vttablet/endtoend) have to bring up the LocalCluster differently.

So, I think we need the following:
* A function to register LocalCluster constructors.
* A factory method that will instantiate a LocalCluster based on the registered function.

Alternately, the above could work for creation of `Environment`, and then `LocalCluster.Setup` would call the factory method. This actually sounds better than the LocalCluster factory.

Once this is done, we can register our constructors which will override the default ones.

sougou commented:
I've managed to rebuild our internal environment under the new vttest, which involved building alternate versions of LocalCluster and MySQLManager. I have one last test that's holding out on me. So, once that's through, we can look at merging this change.

I still need the factory change. Here's how I've currently worked around in local_cluster_old.go:
```
func LaunchVitess(options ...VitessOption) (hdl *Handle, err error) {
      env, _ := newGoogleEnv()
      hdl = &Handle{&LocalCluster{Env:env}}
      err = hdl.run(options...)
      return
}
```
I also have to review the code. The fact that I was able to import successfully means that things look good structurally. So, I may only have a few nits here and there.

vmg commented:
That is great news. I'll fix the factory/constructor issue in a couple hours. 👌

vmg commented:
@sougou: Alright, done. See https://github.com/youtube/vitess/pull/3214/files#diff-a24242678ed0008147113076c729efc8R64

It should now be possible to do:

```go
env, _ := newGoogleEnv()
hdl, err := LaunchVitess(WithEnvironment(env), ProtoTopo(topology), Schema(schema), VSchema(vschema))
```

Does this cover all your use cases?

sougou commented:
This will still require code change on import, right? It would be better if the call to LaunchVitess works as is, but uses newGoogleEnv. What I had in mind was this:
```
func DefaultEnvFactory func(flavor string, basePort int) (Environment, error)

func NewEnv(flavor string, basePort int) (Environment, error) {
	if DefaultEnvFactory == nil {
		return NewLocalTestEnv(flavor, basePort)
	}
	return DefaultEnvFactory(flavor, basePort)
}

func (db *LocalCluster) Setup() error {
	var err error

	if db.Env == nil {
		log.Info("No environment in cluster settings. Creating default...")
		db.Env, err = NewEnv()
		if err != nil {
			return err
		}
	}
	...
}
```

This will allow our internal 'env.go' file to set this function. Or alternately:

```
DefaultEnv Environment

func (db *LocalCluster) Setup() error {
	var err error

	if db.Env == nil {
		log.Info("No environment in cluster settings. Using default...")
		db.Env = DefaultEnv
	}
	...
```

The internal init function will just set DefaultEnv. I think I prefer the second option.

sougou commented:
I fixed some corner cases in my previously pasted code. Now both look equally ugly. So, choose anything, or you can come up with something more elegant :).

vmg commented:
@sougou: What do you think about https://github.com/youtube/vitess/pull/3214/commits/63a836e5cad105931fa751e9e1e36d1c64a62359?

sougou commented:
I think this will work, based on the assumption that static initialization happens before other `init` functions. Let me try to import this.

sougou commented on go/vt/vttest/environment.go:
```diff
@@ -17,34 +18,252 @@ limitations under the License.
+
+// DefaultMySQLFlavor is the MySQL flavor used by vttest when MYSQL_FLAVOR is not
+// set in the environment
+const DefaultMySQLFlavor = "MariaDB"
```
We can change this to MySQL56 since that's the most commonly used flavor right now.

sougou commented on go/cmd/vttestserver/main.go:
```diff
@@ -0,0 +1,214 @@
+package main
```
Our internal linter complained about a missing comment here. Could you add one?

sougou commented on go/vt/vttest/environment.go:
```diff
@@ -17,34 +18,262 @@ limitations under the License.
+
+// TearDown implements TearDown for LocalTestEnv
+func (env *LocalTestEnv) TearDown() error {
+	return os.RemoveAll(env.TmpPath)
```
@alainjobart says that it's better to SIGTERM vtcombo and wait for it to exit. This way, all logs will be flushed, etc.

sougou commented on go/vt/vttest/util.go:
```diff
@@ -0,0 +1,32 @@
+package vttest
```
Let's rename this file to shard_name.go. 'util' has become like 'goto'. You're supposed to avoid it as much as possible :).
